packages/server/modules/workspaces/graph/resolvers/workspaces.ts

```diff
@@ -623,6 +626,34 @@ export = FF_WORKSPACES_MODULE_ENABLED
           return ctx.loaders.workspaces!.getWorkspace.load(args.workspaceId)
         }
       },
+      WorkspaceProjectMutations: {
+        updateRole: async (_parent, args, context) => {
+          const { projectId, userId, role } = args.input
```

Comment:
1. Why do we need a new resolver for project role updates?
2. If the old resolver would not work correctly in this scenario, what prevents an API consumer from using it anyway and introducing some kind of broken state?

Reply:
@gjedlicska If the new resolver is legitimate, then this is the 2nd time already where we need to duplicate a resolver just to inject some workspaces logic in non-workspaces code. The API surface and the implementations are getting complicated (and largely duplicated), so seems to me that figuring out a solution to the "how to dynamically invoke speckle module APIs and get returned values back" is becoming even more relevant.

Reply:
I also want to add that the solution to that problem also would've made workspace invites logic simpler

Reply:
Part of the problem is that we don't know if the project is part of a workspace with just the `ProjectMutations.updateRole` mutation. So we can:

- Introduce these extra fetch/checks on every role update

or 

- Introduce this dedicated path

With Monday brain I'm feeling more okay with adding this overhead to all role changes, so that we can guarantee the protections are in place. Esp so if it helps invites work

Reply:
"Introduce these extra fetch/checks on every role update" not possible if we want to avoid workspace code in non-workspaces modules

Reply:
bubbling up from our convo: investigating using the current event listener/emitter infrastructure for this

Reply:
In general, our graphql API should be the smallest possible wrapper on top of existing services. 

Many times, the usecase for a workspace specific action already has some more business logic to it, that is hard to reasonably inject into a non workspace specific resolver. Duplicating resolvers with re-using already working and sufficiently tested resolvers is not a big maintenance burden compared to how much clarity in implementation it adds both on the backend and the frontend.